### PR TITLE
[Core] Improve event stats for client and server calls

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
- Introduce `RecordEnd()` as an alternative of `RecordExecution()`. `RecordExecution()` is for tracking handler inside an event loop while `RecordEnd()` is for tracking the end of an event manually.
- Track the time from sending the request to receiving the response in client call. Currently we only track the time of `OnReplyReceived()`.
- Track the time from receiving the request to sending the response in server call. Currently we only track the time of `HandleRequestImpl()`.
- Track queueing time per event.

Sample event stats dump:

```
NodeManagerService.grpc_client.RequestWorkerLease - 300 total (0 active), Execution time: mean = 280.820 us, total = 84
.246 ms, Queueing time: mean = 0.000 s, max = -0.000 s, min = 9223372036.855 s, total = 0.000 s
CoreWorker.SubmitTask - 300 total (0 active), Execution time: mean = 63.057 us, total = 18.917 ms, Queueing time: mean 
= 14.177 us, max = 54.000 us, min = 8.000 us, total = 4.253 ms
CoreWorkerService.grpc_client.PushTask.OnReplyReceived - 300 total (0 active), Execution time: mean = 60.217 us, total 
= 18.065 ms, Queueing time: mean = 11.507 us, max = 40.000 us, min = 8.000 us, total = 3.452 ms
```

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/rpc/client_call.h [hunk 1]
```diff
@@ -314,7 +315,11 @@ class ClientCallManager {
                 // The call is finished, and we can delete this tag now.
                 delete tag;
               },
-              std::move(stats_handle));
+              stats_handle->event_name + ".OnReplyReceived",
```
Comment (rkooo567) [inline src/ray/rpc/client_call.h:318]: Is it possible to test this using e2e test + manually calling grpc stub?

**Review by rkooo567 — COMMENTED** [review]
Can you share how we should see the output of event stats to detect leak?  (for actual code review, I will do it tmrw)

File: src/ray/rpc/client_call.h (hunk 1 above)
Comment (jjyao) [reply src/ray/rpc/client_call.h:318]: Added.

**Comment by jjyao** [comment]
> Can you share how we should see the output of event stats to detect leak?

Added a sample event stats dump. By looking at the # active events, we can see if there is leak.

**Comment by rkooo567** [comment]
```
CoreWorkerService.grpc_client.PushTask.OnReplyReceived - 300 total (0 active), Execution time: mean = 57.367 us, total 
= 17.210 ms, Queueing time: mean = 11.280 us, total = 3.384 ms
CoreWorkerService.grpc_client.PushTask - 300 total (0 active), Execution time: mean = 871.657 us, total = 261.497 ms, Q
ueueing time: mean = 0.000 s, total = 0.000 s
```

So active == not replied? 

Or it is CoreWorkerService.grpc_client.PushTask  - CoreWorkerService.grpc_client.PushTask.OnReplyReceived?

**Comment by jjyao** [comment]
Active CoreWorkerService.grpc_client.PushTask == not replied.

File: src/ray/rpc/server_call.h [hunk 2]
```diff
@@ -204,6 +205,7 @@ class ServerCallImpl : public ServerCall {
   void SetState(const ServerCallState &new_state) override { state_ = new_state; }
 
   void HandleRequest() override {
+    stats_handle_ = io_service_.stats().RecordStart(call_name_);
```
Comment (rkooo567) [inline src/ray/rpc/server_call.h:208]: actually, is it possible to track both 

1. handler execution time
2. time to reply

? I think there are times 1 is much more important than 2

File: src/ray/common/event_stats.cc [hunk 3]
```diff
@@ -81,8 +82,26 @@ std::shared_ptr<StatsHandle> EventTracker::RecordStart(
       global_stats_);
 }
 
+void EventTracker::RecordEnd(std::shared_ptr<StatsHandle> handle) {
+  RAY_CHECK(!handle->end_or_execution_recorded);
```
Comment (rkooo567) [inline src/ray/common/event_stats.cc:86]: Should we just make it no-op instead? Seems too hard check for just instrumentation library

File: src/ray/common/event_stats.cc [hunk 4]
```diff
@@ -211,11 +231,15 @@ std::string EventTracker::StatsString() const {
...
     }
-    event_stats_stream << "), CPU time: mean = "
+    event_stats_stream << "), Execution time: mean = "
                        << to_human_readable(entry.second.cum_execution_time /
                                             static_cast<double>(entry.second.cum_count))
                        << ", total = "
-                       << to_human_readable(entry.second.cum_execution_time);
+                       << to_human_readable(entry.second.cum_execution_time)
+                       << ", Queueing time: mean = "
+                       << to_human_readable(entry.second.cum_queue_time /
```
Comment (rkooo567) [inline src/ray/common/event_stats.cc:240]: Q: should we also track the max time here? I feel like max time is actually pretty useful. 

Maybe doesn't have to be included in this PR

File: src/ray/common/client_connection.cc [hunk 5]
```diff
@@ -401,8 +401,8 @@ void ClientConnection::ProcessMessages() {
     auto this_ptr = shared_ClientConnection_from_this();
     auto &io_context = static_cast<instrumented_io_context &>(
         ServerConnection::socket_.get_executor().context());
-    const auto stats_handle =
-        io_context.stats().RecordStart("ClientConnection.async_read.ReadBufferAsync");
+    const auto stats_handle = io_context.stats().RecordStart(
+        "ClientConnection.async_read.ProcessMessageHeader");
```
Comment (rkooo567) [inline src/ray/common/client_connection.cc:405]: nice discovery!

File: src/ray/rpc/server_call.h (hunk 2 above)
Comment (jjyao) [reply src/ray/rpc/server_call.h:208]: We already have 1.

```
if (!io_service_.stopped()) {
      io_service_.post([this, auth_success] { HandleRequestImpl(auth_success); },
                       call_name_ + ".HandleRequestImpl",
```

File: src/ray/common/event_stats.cc (hunk 4 above)
Comment (jjyao) [reply src/ray/common/event_stats.cc:240]: Good point. Let me add in this PR.

File: src/ray/common/event_stats.cc (hunk 3 above)
Comment (jjyao) [reply src/ray/common/event_stats.cc:86]: Added a comment in the docstring saying the pair of methods must be called.

**Review by rkooo567 — APPROVED** [review]
Lgtm!

**Pull request merged**